ext/node/global.rs

```diff
@@ -0,0 +1,448 @@
+  let reflect = global
+    .get(scope, reflect_key.into())
+    .unwrap()
+    .to_object(scope)
+    .unwrap();
+  let get_key = v8::String::new_external_onebyte_static(scope, b"get").unwrap();
+  let reflect_get = reflect.get(scope, get_key.into()).unwrap();
+  assert_eq!(reflect_get.is_function(), true);
+  let set_key = v8::String::new_external_onebyte_static(scope, b"set").unwrap();
+  let reflect_set = reflect.get(scope, set_key.into()).unwrap();
+  assert_eq!(reflect_set.is_function(), true);
+
+  // globalThis.__bootstrap.ext_node_denoGlobals and
+  // globalThis.__bootstrap.ext_node_nodeGlobals are the objects that contain
+  // the Deno and Node specific globals respectively. If they do not yet exist
+  // on the global object, create them as null prototype objects.
+  let bootstrap_key =
+    v8::String::new_external_onebyte_static(scope, b"__bootstrap").unwrap();
+  let bootstrap = match global.get(scope, bootstrap_key.into()) {
+    Some(value) if value.is_object() => value.to_object(scope).unwrap(),
+    Some(value) if value.is_undefined() => {
+      let null = v8::null(scope);
+      let obj =
+        v8::Object::with_prototype_and_properties(scope, null.into(), &[], &[]);
+      global.set(scope, bootstrap_key.into(), obj.into());
+      obj
+    }
+    _ => panic!("__bootstrap should not be tampered with"),
+  };
+  let deno_globals_key =
+    v8::String::new_external_onebyte_static(scope, b"ext_node_denoGlobals")
+      .unwrap();
+  let deno_globals = match bootstrap.get(scope, deno_globals_key.into()) {
+    Some(value) if value.is_object() => value,
+    Some(value) if value.is_undefined() => {
+      let null = v8::null(scope);
+      let obj =
+        v8::Object::with_prototype_and_properties(scope, null.into(), &[], &[])
+          .into();
+      bootstrap.set(scope, deno_globals_key.into(), obj);
+      obj
+    }
+    _ => panic!("__bootstrap.ext_node_denoGlobals should not be tampered with"),
+  };
+  let node_globals_key =
+    v8::String::new_external_onebyte_static(scope, b"ext_node_nodeGlobals")
+      .unwrap();
+  let node_globals = match bootstrap.get(scope, node_globals_key.into()) {
+    Some(value) if value.is_object() => value,
+    Some(value) if value.is_undefined() => {
+      let null = v8::null(scope);
+      let obj =
+        v8::Object::with_prototype_and_properties(scope, null.into(), &[], &[])
+          .into();
+      bootstrap.set(scope, node_globals_key.into(), obj);
+      obj
+    }
+    _ => panic!("__bootstrap.ext_node_nodeGlobals should not be tampered with"),
+  };
+
+  // set the internal fields
+  assert!(global.set_internal_field(0, reflect_get));
+  assert!(global.set_internal_field(1, reflect_set));
+  assert!(global.set_internal_field(2, deno_globals));
+  assert!(global.set_internal_field(3, node_globals));
+}
+
+fn is_managed_key(
+  scope: &mut v8::HandleScope,
+  key: v8::Local<v8::Name>,
+) -> bool {
+  let Ok(str): Result<v8::Local<v8::String>, _> = key.try_into() else {
```

Comment:
If this is expected to be called a lot, it should probably fast-case https://docs.rs/rusty_v8/latest/rusty_v8/struct.String.html#method.is_onebyte and use it to write a one-byte string that can be compared much faster.

Reply:
According to the v8 docs, `IsOneByte` can have false negatives rather than false positives -- the comment in rusty_v8 might be wrong. https://v8.github.io/api/head/classv8_1_1String.html#a8f14ab3aff52295d2d3245081a1b29eb

Reply:
Regarding string copies and comparisons, it's indeed probable that OneByte doesn't actually that significantly change the performance game: it's after all a malloc before the comparison.

The optimal situation would be to use static OneByteExternals and create locals out of those for comparisons. This is also what serde_v8 should be doing. Elsewhere I've also had code setup to always externalize V8 strings that equal   an external string. When the externalized string is coming from eg. a WebSocket it's wasted time but when it's a JS side constant it becomes a great boon: Future comparisons of this string with the externals becomes a pointer comparison.

Reply:
Revisiting after spending some time deep in ops. I've gone and fixed the `IsOneByte` documentation issues in the rusty_v8 repo.

We don't currently have any sort of benchmarks on any of this code. I think it's a bad idea to land the code without any understanding of what impact it might have on global object access timings, especially considering the fact that this code could be running in hot loops inside `ws` or other important packages. V8 works in surprising ways, proxies may not land in the optimial paths for JIT, and something like this should probably be compared against the existing system to ensure that we aren't regressing anything.

Reply:
@mmastrac please read the pr again. The proxy is never used for real properties of the global object (real properties shadow the proxy). I have benchmarked this case, and there is no measurable slowdown for real global property access.

Reply:
The proxy is only called on set of new properties, and when accessing properties that don't exist (ie that in strict mode would result in a ReferenceError.